### PR TITLE
Add Check for Token Issuer and Expiration

### PR DESCRIPTION
**Pull request by charlie-pisuraj**
Makes the DSS reject token with missing Expiration and Issuer field #127 
In addition the standard mentions that the token cannot be valid for more than 1 hour into the future so we also check it and reject token are are given too wide of a window

### CODE REVIEW

File: cmds/dummy-oauth/main.go [hunk 1]
```diff
@@ -46,9 +48,29 @@ func createGetTokenHandler(privateKey *rsa.PrivateKey) http.Handler {
...
 		}
+
+		var (
+			issuer  string   = ""
+			issuers []string = params["issuer"]
+		)
+		if len(issuers) == 1 {
+			issuer = issuers[0]
+		}
+		var (
```
Comment (steeling) [inline cmds/dummy-oauth/main.go]: let's put all of the var blocks at the top of this function. it's getting a little long, and there's no early return so no extra cost in creating them

File: pkg/dss/auth/auth.go [hunk 2]
```diff
@@ -42,6 +42,14 @@ func (m *missingScopesError) Error() string {
 	return strings.Join(m.s, ", ")
 }
 
+type tokenExpireError struct {
```
Comment (steeling) [inline pkg/dss/auth/auth.go]: nit: how about 2 vars: tokenExpired, and tokenExpiredTooFar, ie:

tokenExpired = errors.New("the token is already expired")

File: pkg/dss/auth/auth.go [hunk 3]
```diff
@@ -229,9 +237,31 @@ func (a *Authorizer) AuthInterceptor(ctx context.Context, req interface{}, info
...
 
+func (a *Authorizer) checkExpired(claimedExpireTime int64) error {
+	now := time.Now()
+	if claimedExpireTime < now.Unix() {
+		return &tokenExpireError{msg: "Token Expired or Expiration missing"}
+	}
+
+	if claimedExpireTime > now.Add(time.Hour).Unix() {
+		return &tokenExpireError{
+			msg: "Token expiration time is too far in the furture, Max token duration is 1 Hour",
```
Comment (steeling) [inline pkg/dss/auth/auth.go]: @BenjaminPelletier can you comment on this piece with respect to the standards?

Comment (BenjaminPelletier) [reply pkg/dss/auth/auth.go]: One hour in the future is from the YAML text for AuthFromAuthorizationAuthority describing required fields in the JWT: "exp, with a time no further than 1 hour in the future."  So, this looks like a good test to me.

**Comment by charlie-pisuraj** [comment]
All comments resolved.

File: pkg/dss/auth/auth.go [hunk 4]
```diff
@@ -229,9 +247,32 @@ func (a *Authorizer) AuthInterceptor(ctx context.Context, req interface{}, info
...
+	}
+
+	if err := a.checkExpired(claims.ExpiresAt); err != nil {
+		return nil, dsserr.Unauthenticated(fmt.Sprintf("%v", err))
+	}
+
 	return handler(ContextWithOwner(ctx, models.Owner(claims.Subject)), req)
 }
 
+func (a *Authorizer) checkExpired(claimedExpireTime int64) error {
```
Comment (steeling) [inline pkg/dss/auth/auth.go]: Can we add the checkExpired to the claims.Valid() ?
https://github.com/interuss/dss/blob/60d63fb8dc0eee095d688d4d9427961dfbce15f1/pkg/dss/auth/claims.go#L38

This is how it is done in jwt.Valid, so it makes sense to put it there.

Also a nit: we can use their expiredAt if you choose too https://github.com/dgrijalva/jwt-go/blob/master/claims.go#L69

**Comment by steeling** [comment]
LGTM

**Comment by charlie-pisuraj** [comment]
I don't understand whats failing, I ran the test locally and it all passed
the logs pasted is just the logging error we expect from it.

**Pull request merged**